include/net/dsa.h

```diff
@@ -0,0 +1,174 @@
+ * master (e.g. implement tail tagging).
+ *
+ * @param dev Device
+ * @param pkt Network packet
+ *
+ * Returns:
+ *  - Pointer to (modified) net_pkt structure
+ */
+struct net_pkt *dsa_xmit_pkt(struct device *dev, struct net_pkt *pkt);
+
+/**
+ * @brief DSA generic transmit function
+ *
+ * This is a generic function for passing packets from slave DSA interface to
+ * master.
+ *
+ * @param dev Device
+ * @param pkt Network packet
+ *
+ * Returns:
+ *  - 0 if ok (packet sent via master iface), < 0 if error
+ */
+int dsa_tx(struct device *dev, struct net_pkt *pkt);
+
+
+/**
+ * @brief DSA function to get proper interface
+ *
+ * This is a generic function for assigning proper slave interface after
+ * receiving the packet on master.
+ *
+ * @param iface Network interface
+ * @param pkt Network packet
+ *
+ * Returns:
+ *  - Pointer to struct net_if
+ */
+struct net_if *dsa_get_iface(struct net_if *iface, struct net_pkt *pkt);
+
+/**
+ * @brief DSA (MGMT) Receive packet callback
+ *
+ * Callback gets called upon receiving packet. It is responsible for
+ * freeing packet or indicating to the stack that it needs to free packet
+ * by returning correct net_verdict.
+ *
+ * Returns:
+ *  - NET_DROP, if packet was invalid, rejected or we want the stack to free it.
+ *    In this case the core stack will free the packet.
+ *  - NET_OK, if the packet was accepted, in this case the ownership of the
+ *    net_pkt goes to callback and core network stack will forget it.
+ */
+typedef enum net_verdict (*net_dsa_recv_cb_t)(struct net_if *iface,
+					      struct net_pkt *pkt);
+
+/**
+ * @brief Register DSA Rx callback functions
+ *
+ * @param iface Network interface
+ * @param cbrx Receive callback function
+ *
+ * @return 0 if ok, < 0 if error
+ */
+int dsa_register_recv_callback(struct net_if *iface, net_dsa_recv_cb_t cb);
+
+/**
+ * @brief Parse DSA packet
+ *
+ * @param iface Network interface (master)
+ * @param pkt Network packet
+ *
+ * @return Return NET_DROP if packet was invalid, rejected or we want the stack
+ *         to free it. otherwise NET_OK
+ */
+enum net_verdict net_dsa_recv(struct net_if *iface, struct net_pkt *pkt);
+
+/**
+ * @brief Set DSA interface to packet
+ *
+ * @param iface Network interface (master)
+ * @param pkt Network packet
+ *
+ * @return Return the slave network interface
+ */
+struct net_if *dsa_recv_set_iface(struct net_if *iface, struct net_pkt **pkt);
+
+/**
+ * @brief Write value to DSA register
+ */
+int dsa_write_reg(u16_t reg_addr, u8_t value);
+
+/**
+ * @brief Read value from DSA register
+ */
+int dsa_read_reg(u16_t reg_addr, u8_t *value);
+
+/**
+ * @brief Enable DSA port
+ *
+ * @param iface Network interface (master)
+ * @param port Port number to be enabled
+ *
+ * @return 0 if ok, < 0 if error
+ */
+int dsa_enable_port(struct net_if *iface, u8_t port);
+
+/**
+ * @brief Structure to provide dsa context
+ */
+
+struct dsa_context {
```

Comment:
This seems to be driver specific data so should not be defined in public header.